### DISCUSSION TITLE
Support older build targets

### DISCUSSION THREAD

**Discussion by milamer** [remix-run/remix#2557] [Proposals]
ESBuild supports [different build targets](https://esbuild.github.io/content-types/#javascript) to make the code backwards compatible with older devices. Currently remix only provides a [target](https://esbuild.github.io/api/#target) in the build command for the server build, but not the client build.

See [compiler.ts](https://github.com/remix-run/remix/blob/9021fab2aa727ea423ed089550998227796f777c/packages/remix-dev/compiler.ts#L213)

**Comment by milamer** [comment]
### Patch

For `0.18.0-pre.0`
```patch
diff --git a/node_modules/@remix-run/dev/compiler.js b/node_modules/@remix-run/dev/compiler.js
index 923d122..e9378a7 100644
--- a/node_modules/@remix-run/dev/compiler.js
+++ b/node_modules/@remix-run/dev/compiler.js
@@ -177,6 +177,7 @@ async function createBrowserBuild(config, options) {
   return esbuild.build({
     entryPoints,
     outdir: config.assetsBuildDirectory,
+    target: config.browserBuildTarget,
     platform: "browser",
     format: "esm",
     external: externals,
diff --git a/node_modules/@remix-run/dev/config.js b/node_modules/@remix-run/dev/config.js
index e4d5d8a..9f3b85e 100644
--- a/node_modules/@remix-run/dev/config.js
+++ b/node_modules/@remix-run/dev/config.js
@@ -100,6 +100,7 @@ async function readConfig(remixRoot, serverMode = serverModes.ServerMode.Product
     devServerPort,
     assetsBuildDirectory,
     publicPath,
+    browserBuildTarget: appConfig.browserBuildTarget,
     rootDirectory,
     routes: routes$1,
     serverBuildDirectory,
```

With this you can add a build target in `remix.config.js`
```js
module.exports = {
  appDirectory: 'app',
  browserBuildDirectory: 'public/build',
  browserBuildTarget: 'es2018',
  publicPath: '/build/',
  serverBuildDirectory: 'build',
  devServerPort: 8002,
}
```

### Supported features

Be aware that unsupported features can lead to endless loops (e.g. build target `es5` with `const` in you code).

> Note that if you use a syntax feature that esbuild doesn't yet have support for transforming to your current language target, esbuild will generate an error where the unsupported syntax is used. This is often the case when targeting the es5 language version, for example, since esbuild only supports transforming most newer JavaScript syntax features to es6.

More info on supported [features](https://esbuild.github.io/content-types/#javascript)

### Build targets

You can define the build target in the `remix.config.js` as `browserBuildTarget`. It can either be a single string with the desired target (e.g. `es2018`) or an array of targets. It also supports browser versions!

```ts
interface Config {
  browserBuildTarget : string | Array<string>
}
// eg
const buildTarget = [
    'es2020',
    'chrome58',
    'firefox57',
    'safari11',
    'edge16',
    'node12',
]
```

More info on possible [build targets](https://esbuild.github.io/api/#target)

**Comment by markiavelli7** [comment]
Thanks @milamer!! You rock!

The structure of @remix-run/dev/config.js has changed a little bit in v0.18+, but it's a pretty minor change and the same principle of declaring and returning the build target is the same.

**Comment by vijaybritto** [comment]
I'm trying to port part of our Enterprise app to Remix and the browser support we need is ES2018 at least. Is adding `target` property going to be supported any time? Or there any technical reasons for this?
I can open a PR.

**Comment by ghost** [comment]
Need it too.

**Comment by machour** [comment]
@vijaybritto we always welcome PRs :)

**Comment by sergiodxa** [comment]
> You can define the build target in the remix.config.js as browserBuildTarget. It can either be a single string with the desired target (e.g. es2018) or an array of targets. It also supports browser versions!

I think using [.browserlist](https://github.com/browserslist/browserslist) for this would be better since that's also used by other tools.

**Comment by mciparelli** [comment]
It's currently not possible to specify a build target and it defaults to node 14 (https://github.com/remix-run/remix/blob/main/packages/remix-dev/build.ts#L19)
top level await does not work because of this.

**Comment by soundstep** [comment]
I've got similar needs I think, both for the js client output and also the js that is in the html page.
I've created a [discussion in the Remix discord](https://discord.com/channels/770287896669978684/1151143089302413361/1151143089302413361) to get some advices.
